waku/v2/node/dnsdisc/waku_dnsdisc.nim

```diff
@@ -0,0 +1,171 @@
+    return some(IpTransportProtocol.tcpProtocol)
+
+  if typedR.udp6.isSome or typedR.udp.isSome:
+    return some(IpTransportProtocol.udpProtocol)
+
+  return none(IpTransportProtocol)
+
+func toPeerInfo*(enr: enr.Record): Result[PeerInfo, cstring] =
+  let typedR = ? enr.toTypedRecord
+
+  if not typedR.secp256k1.isSome:
+    return err("enr: no secp256k1 key in record")
+  
+  let
+    pubKey = ? keys.PublicKey.fromRaw(typedR.secp256k1.get)
+    peerId = ? PeerID.init(crypto.PublicKey(scheme: Secp256k1,
+                                            skkey: secp.SkPublicKey(pubKey)))
+  
+  var addrs = newSeq[MultiAddress]()
+
+  let transportProto = getTransportProtocol(typedR)
+  if transportProto.isNone:
+    return err("enr: could not determine transport protocol")
+
+  case transportProto.get()
+  of tcpProtocol:
+    if typedR.ip.isSome and typedR.tcp.isSome:
+      let ip = ipv4(typedR.ip.get)
+      addrs.add MultiAddress.init(ip, tcpProtocol, Port typedR.tcp.get)
+
+    if typedR.ip6.isSome:
+      let ip = ipv6(typedR.ip6.get)
+      if typedR.tcp6.isSome:
+        addrs.add MultiAddress.init(ip, tcpProtocol, Port typedR.tcp6.get)
+      elif typedR.tcp.isSome:
+        addrs.add MultiAddress.init(ip, tcpProtocol, Port typedR.tcp.get)
+      else:
+        discard
+
+  of udpProtocol:
+    if typedR.ip.isSome and typedR.udp.isSome:
+      let ip = ipv4(typedR.ip.get)
+      addrs.add MultiAddress.init(ip, udpProtocol, Port typedR.udp.get)
+
+    if typedR.ip6.isSome:
+      let ip = ipv6(typedR.ip6.get)
+      if typedR.udp6.isSome:
+        addrs.add MultiAddress.init(ip, udpProtocol, Port typedR.udp6.get)
+      elif typedR.udp.isSome:
+        addrs.add MultiAddress.init(ip, udpProtocol, Port typedR.udp.get)
+      else:
+        discard
+
+  if addrs.len == 0:
+    return err("enr: no addresses in record")
+
+  return ok(PeerInfo.init(peerId, addrs))
+
+func createEnr*(privateKey: crypto.PrivateKey,
+                enrIp: Option[ValidIpAddress],
+                enrTcpPort, enrUdpPort: Option[Port]): enr.Record =
+  
+  assert privateKey.scheme == PKScheme.Secp256k1
+
+  let
+    rawPk = privateKey.getRawBytes().expect("Private key is valid")
+    pk = keys.PrivateKey.fromRaw(rawPk).expect("Raw private key is of valid length")
+    enr = enr.Record.init(1, pk, enrIp, enrTcpPort, enrUdpPort).expect("Record within size limits")
+  
+  return enr
+
+#####################
+# DNS Discovery API #
+#####################
+
+proc emptyResolver*(domain: string): Future[string] {.async, gcsafe.} =
+  return ""
+
+proc findPeers*(wdd: var WakuDnsDiscovery): Result[seq[PeerInfo], cstring] =
+  ## Find peers to connect to using DNS based discovery
+  
+  info "Finding peers using Waku DNS discovery"
+  
+  # Synchronise client tree using configured resolver
+  var tree: Tree
+  try:
+    tree = wdd.client.getTree(wdd.resolver)  # @TODO: this is currently a blocking operation to not violate memory safety
```

Comment:
Is there a timeout?

Reply:
Yes, there is an [underlying timeout for the DNS resolution](https://github.com/status-im/nim-dnsdisc/blob/dcb9290d004476fb0a5389baa88121b072abf135/discovery/dnsdisc/client.nim#L37).
Currently hard-coded to `20.seconds`, though this can of course be configured in future.